File: backend/src/main/java/com/github/davidpotentini/model/modelos/ModeloModel.java
package com.github.davidpotentini.model.modelos;

import com.github.davidpotentini.enums.EPeriodicidade;
import com.github.davidpotentini.enums.EStatusModelo;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDateTime;

/**
 * Modelo de planejamento — herda a estrutura de processos/práticas da metodologia da incubadora (não
 * copia — só lê ao vivo); o que o modelo guarda são as {@link AtividadeModeloModel atividades} de cada
 * prática.
 *
 * <p>Enquanto {@code RASCUNHO} as atividades podem ser editadas; ao {@code PUBLICADO} o modelo fica
 * imutável. Schema do tenant.
 */
@Entity
@Table(name = "MODELOS")
@Getter
@Setter
public class ModeloModel {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "MOD_COD")
    private Long modCod;

    @Column(name = "NOME", nullable = false)
    private String nome;

    @Enumerated(EnumType.STRING)
    @Column(name = "PERIODICIDADE")
    private EPeriodicidade periodicidade = EPeriodicidade.ANUAL;

    @Enumerated(EnumType.STRING)
    @Column(name = "STATUS", nullable = false)
    private EStatusModelo status = EStatusModelo.RASCUNHO;

    /** Preenchido na publicação; {@code null} enquanto rascunho. */
    @Column(name = "PUBLICADO_EM")
    private LocalDateTime publicadoEm;

    @Column(name = "DESCRICAO")
    private String descricao;
}

File: frontend/src/app/models/modelos/modelo.model.ts
import { EAtivoInativo, EPeriodicidade } from '../metodologia/metodologia.model';

export type EStatusModelo = 'RASCUNHO' | 'PUBLICADO';

export const STATUS_MODELO_LABEL: Record<EStatusModelo, string> = {
  RASCUNHO: 'Rascunho',
  PUBLICADO: 'Publicado',
};

/**
 * Modelo de planejamento. Herda a estrutura da metodologia da incubadora (lida ao vivo). Enquanto
 * `RASCUNHO` as atividades são editáveis; `PUBLICADO` é imutável.
 */
export interface Modelo {
  modCod: number;
  nome: string;
  periodicidade: EPeriodicidade;
  status: EStatusModelo;
  publicadoEm: string | null;
  descricao: string | null;
}

/** Atividade de uma prática do modelo (único nível editável na tela). */
export interface AtividadeModelo {
  atmCod: number;
  modCod: number;
  prtCod: number;
  nome: string;
  observacoes: string | null;
  respPesCod: number | null;
  situacao: EAtivoInativo;
}

/** Prática dentro da estrutura do modelo — só leitura (da metodologia) + atividades. */
export interface ModeloPratica {
  prtCod: number;
  nome: string;
  descricao: string | null;
  atividades: AtividadeModelo[];
}

/** Processo dentro da estrutura do modelo — só leitura, ordenado por `ordem`. */
export interface ModeloProcesso {
  prcCod: number;
  ordem: number;
  nome: string;
  descricao: string | null;
  praticas: ModeloPratica[];
}
